### DISCUSSION TITLE
[Draft Issue]: Anonymous Interface Implementation (new interface anonymously)

### DISCUSSION THREAD

**Discussion by reitowo** [dotnet/csharplang#5818] [Language Ideas]
I forgot the formal name of this feature, please remind me.

One of my Java's favourite feature is:
```java
runOnUiThread(new Runnable() {
    void run() {
        // Code here
    }
});
```
In case you say just use an lambda, another example: 
```java
client.perform(new Resp<Data>() {
    void onSucceed(Data data) {
        // Code here
    }
    void onFailed(ErrorCode error) {
        // Code here
    }
});
```
where `Resp` is an interface.

It implements an interface in an anonymous way, which creates an anonymous class (type). We don't need to implement some interface in an extra file anymore, just create in function.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/2517

**Reply by reitowo** [reply]
The anonymous interface implementation is a great Java feature that I have often missed in the C#/.NET land. I believe that this is a better solution than a single use classes that either pollute the namespace or need to be in some way hidden from view.

**Reply by reitowo** [reply]
I don't think it is a duplicate and it's usage is not only about interop with Java. It's about easy coding. There're many times that we just don't want to use a separate class to implement interface, and just want to declare a anonymous one in the function

**Reply by HaloFour** [reply]
My discussion isn't only about the interop with Java either, but the language team has so far been pretty uninterested in exploring anonymous implementations:

https://github.com/dotnet/roslyn/issues/13#issuecomment-173413281

I brought it up in the context of Java interop as hopefully an important use case that currently requires a lot of unnecessary boilerplate and that such interop might be considered worthwhile enough to justify the feature in general.

**Reply by reitowo** [reply]
It was in 2016 and I think from C#8 the design team has changed a lot. So team, please do reconsider this feature.

**Reply by HaloFour** [reply]
Would be nice if they did, I would like having this feature added to the language.

Note that a number of C# designers came from Java, including the lead designer, so they're quite familiar with the feature.  Although interestingly gafter, who is no longer on the team, believed that local classes were more common, and that's not my experience.

Honestly, I'd probably be happy if C# would support lambdas with SAM types as well as delegates.

**Comment by reitowo** [comment]
I think this should be the real big syntax candy that should be in C# 11/12 right away.
This real relieve some pain about creating unnecessary classes. And we also want function scoped types too.

**Comment by huoyaoyuan** [comment]
There is already a proposal #4301
